### ISSUE TITLE
How to determine if there are active subscriptions?

### ISSUE THREAD

**Issue by TepeoIain** [aws/aws-appsync-community#129]
We want to display the current state of our IoT devices in our Angular front-end, connected via AppSync subscriptions. When we get a device shadow update from our devices via IoT Core we use a Lambda to call a mutation in AppSync with a local resolver. This publishes notifications to any subscribers, which triggers a front-end update. That's all good.

However this does mean that sometimes our Lambda is calling mutations in AppSync when there are no subscribers to be notified, either because there are no clients connected at that time, or because the connected clients are not displaying data from that particular IoT device. We really don't want to do this at scale, as it would be quite expensive.

What would be great is a way for the Lambda to get AppSync to confirm whether there is a subscription for a particular mutation and IoT device prior to calling the mutation. Is there a way to do this?

**Comment by gsans** [comment]
I think you may be able to achieve this behaviour by filtering by IoT device Id. This would allow you to get only updates for a specific device or all when no Id is provided; and unsubscribe if you are not interested in neither.

Resources:
https://docs.amplify.aws/guides/api-graphql/subscriptions-by-id/q/platform/js

https://gist.github.com/gsans/aeffbae886add35b669c84f5e5b3fa69

**Comment by TepeoIain** [comment]
Hi Gerard. Thanks for this. We are using this approach from the front-end client, i.e. subscribing only to IoT device IDs that we want to get notifications from. What we aren't able to do from the back-end is to detect when there are no front-end clients subscribed to a IoT device ID, so we can avoid calling the mutation and thus incurring cost.

**Comment by gsans** [comment]
Correct. What costs are you referring to? As far as I understand a subscription with no subscribers will send no notifications. Or do you mean the mutation itself bc is frequent?

**Comment by TepeoIain** [comment]
Yes, we only pay the $2/million subscription notifications when there are front-end clients connected and subscribed. But we are paying $4/million mutations, whether that triggers a subscription notification or not.

**Comment by gsans** [comment]
I've look around a bit and it may be possible using ActiveSubscription (Sum) metric in CloudWatch.

(Announcement)[https://aws.amazon.com/blogs/mobile/appsync-realtime/]
(Monitoring and Logging)[https://docs.aws.amazon.com/appsync/latest/devguide/monitoring.html#cw-metrics]

So the idea would be to:
1) Create a CloudWatch Alarm that looks at AppSync.ActiveSubscription (Sum) <= 0 (every 5min)
2) Whenever the alarm is raised setup to send a notification to an Amazon SNS topic (Eg: appsync-active-subsriptions connected to a Lambda of your choice)

Also notice that you could also use the opposite so you only run the mutation as a result of an alarm that looks that there are AppSync.ActiveSubscription (Sum) > 0 with a sensible schedule.

**Comment by TepeoIain** [comment]
Yes, we spotted that metric in CloudWatch. It does tell us if there aren't any connected clients, or connected clients without subscriptions. So we definitely shouldn't be calling the mutation in that circumstance.

But if we have do have a connected client with a subscription for IoT device ID 001, we don't want to be calling mutations for data coming from IoT device ID 002, and so on.

Our current workaround proposal is to have a Dynamo table where we keep a tally of subscriptions against each IoT device ID. Every client would have to call a mutation on that table when subscribing, and again when unsubscribing. But it would let our Lambda do a quick check when receiving new data from IoT Core before calling the mutation in AppSync to notify subscribers. We were just hoping that AppSync would expose that information directly.

**Comment by gsans** [comment]
Right. That would be my next proposal.  What is the context of this mutation? I don't fully understand how you are happy to lose the information coming from that mutation if no devices are listening.

**Comment by TepeoIain** [comment]
In this context our front-end is providing a 'near' real-time view of the state of our IoT device(s). For example we have a temperature sensor in the device, so when the temperature changes we update the device shadow, which is sync-ed with IoT Core, which triggers the Lambda, which calls the mutation, and any client subscribed to that device then updates the displayed temperature value in the front-end. If there are no clients subscribed then we don't mind 'losing' that data, as there's no-one interested in seeing it.

The mutation has a local resolver, as we're not using the mutation to put the data into Dynamo or Aurora. We could, but don't due to cost. Instead we are storing the data from our IoT devices in RDS via Lambda.

**Comment by gsans** [comment]
Ok. Thanks. I understand better now

**Comment by kavinduraisamy** [comment]
Any update on this ? We are also looking this feature

**Comment by daklimek** [comment]
I'm also interested!
